### ISSUE TITLE
Optimisations with Observable.empty()

### ISSUE THREAD

**Issue by davidmoten** [ReactiveX/RxJava#1653]
I notice that calling ```observable.concat(Observable.empty())``` invokes back-pressure. I'd find it useful if when I used ```Observable.empty()``` that optimisations came into play such as ensuring that ```observable.concat(Observable.empty())``` simply returned ```observable```.

To achieve this I'd make a private constant ```EMPTY = from(new ArrayList())```  in ```Observable``` so that calling ```Observable.empty()``` always returned this object cast into the appropriate generic type. Then I'd use a simple object reference equality test to determine if an optimisation could be made.

Do you think this proposal has legs? If so, I'll knock up a PR. I would seek to optimise more than just the ```concat``` operator , it would include ```merge``` and possibly others which I can think about if the time comes.

**Comment by akarnokd** [comment]
Why do you have so many ```empty()``` to merge?

**Comment by davidmoten** [comment]
Ta for improvement. The reason I have ```empty()```s to merge is that this is a common pattern for me:

```java
o.flatMap(t -> {
   if (cond(t)) 
       return someObservable(t);
  else 
       return Observable.empty(); 
  });
```
I'm experimenting now with substituting the special case where the flatMap function returns either 0 or 1 by doing 

```java
o.map(toOptional).filter(isPresent).map(toValue);
```
I'd hope it would be faster but not sure.

**Comment by akarnokd** [comment]
This PR is quite old. Can you create a reasonable benchmark to see how concatMap and flatMap behaves? In addition, I suggest we tackle this after both #2928 and #2960 are merged.

**Comment by davidmoten** [comment]
Yep I'll do benchmarks after those PRs have been merged.

**Comment by akarnokd** [comment]
This may be part of the operator-fusion optimizations in 2.x but I don't see this in 1.x due to non-technical reasons.

**Comment by akarnokd** [comment]
For a start: here is the baseline perf: #3754. I'm going to merge it so subsequent PRs can be benchmarked with it.

**Comment by akarnokd** [comment]
See #3759 for the optimized `concat`s.

**Comment by akarnokd** [comment]
See #3761 for the optimized `merge`s

**Comment by akarnokd** [comment]
Optimizations have been added and released with 1.1.4. Closing the issue. If you have more ideas where it's worth adding extra checks for empty, don't hesitate to open a new issue.
